### PR TITLE
feat: Add CLI commands for managing Agent Skills

### PR DESCRIPTION
**Pull request by uzaxirr**
## Summary

  Adds ag skill CLI commands for creating, validating, and listing Agent Skills.

  - ag skill create - Scaffolds a new skill directory with SKILL.md, scripts/, and references/ subdirectories (scripts and references are created by default)
  - ag skill validate - Validates a skill directory against the Agent Skills specification
  - ag skill list - Lists all valid skills in a directory with optional verbose table view

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Breaking change
- [ ] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [ ] Code complies with style guidelines
- [ ] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [ ] Tested in clean environment
- [ ] Tests added/updated (if applicable)

---

## Additional Notes

Add any important context (deployment instructions, screenshots, security considerations, etc.)

### CODE REVIEW

File: libs/agno_infra/agno/skill/operator.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+        console.print("[bold]Next steps:[/bold]")
+        console.print(f"  1. Edit [cyan]{skill_dir}/SKILL.md[/cyan] to add instructions")
+        if with_scripts:
+            console.print(f"  2. Add scripts to [cyan]{skill_dir}/scripts/[/cyan]")
+        if with_references:
+            console.print(
+                f"  {'3' if with_scripts else '2'}. Add reference docs to [cyan]{skill_dir}/references/[/cyan]"
+            )
+        console.print(
+            f"  {'4' if with_scripts and with_references else '3' if with_scripts or with_references else '2'}. Run [cyan]ag skill validate {skill_dir}[/cyan] to verify"
```
[the remark this answers is not in the document]
Comment (kausmeows) [reply libs/agno_infra/agno/skill/operator.py]: this is valid.

File: libs/agno_infra/agno/skill/operator.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+        errors.append("Skill name can only contain letters, digits, and hyphens")
+
+    return errors
+
+
+def create_skill_from_template(
+    name: Optional[str] = None,
+    path: Optional[str] = None,
+    description: Optional[str] = None,
+    with_scripts: bool = False,
```
Comment (kausmeows) [inline libs/agno_infra/agno/skill/operator.py]: These are `False` by default but the Typer doc says `(created by default)`?
```
  scripts/      - Executable scripts the agent can run (created by default)
  references/   - Reference docs the agent can load on-demand (created by default)
```

File: libs/agno_infra/agno/skill/operator.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+        console.print(f"\n[bold red]Found {len(errors_found)} invalid skill(s):[/bold red]\n")
+        for skill_name, errors in errors_found:
+            console.print(f"  [bold]{skill_name}[/bold]:", style=error_style)
+            for error in errors[:3]:  # Show first 3 errors
+                console.print(f"    - {error}")
+            if len(errors) > 3:
+                console.print(f"    ... and {len(errors) - 3} more error(s)")
+
+
+def _load_skill_info(skill_path: Path) -> Tuple[bool, Any]:
```
Comment (kausmeows) [inline libs/agno_infra/agno/skill/operator.py]: Does the return type have to contain Any? Would be good to type it if we exactly know what to return

File: libs/agno_infra/agno/skill/operator.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+    """
+    errors = []
+
+    if not name or not name.strip():
+        errors.append("Skill name cannot be empty")
+        return errors
+
+    name = name.strip()
+
+    if len(name) > 64:
```
Comment (kausmeows) [inline libs/agno_infra/agno/skill/operator.py]: should use `MAX_SKILL_NAME_LENGTH ` here

File: libs/agno_infra/agno/skill/operator.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+    # Get description interactively if not provided
+    if description is None:
+        description = Prompt.ask(
+            "Short description",
+            default="A skill for agents",
+            console=console,
+        )
+
+    if len(description) > 1024:
```
Comment (kausmeows) [inline libs/agno_infra/agno/skill/operator.py]: why 1024 exactly?

**Review by kausmeows — COMMENTED** [review]
Also will the `ag skill create` give executable scripts? or would users need to make them exec manually?

File: libs/agno_infra/agno/skill/operator.py (hunk 5 above)
Comment (uzaxirr) [reply libs/agno_infra/agno/skill/operator.py]: Its the limit that we previously agreed upon for specs: https://github.com/agno-agi/agno/blob/main/libs/agno/agno/skills/validator.py#L9

File: libs/agno_infra/agno/skill/operator.py (hunk 3 above)
Comment (uzaxirr) [reply libs/agno_infra/agno/skill/operator.py]: added typed

**Comment by uzaxirr** [comment]
> Also will the `ag skill create` give executable scripts? or would users need to make them exec manually?

it does not set executable permissions - user will have to manually do chmod

File: libs/agno_infra/agno/skill/operator.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+    print_info,
+    print_subheading,
+    success_style,
+)
+from agno.skills.validator import MAX_DESCRIPTION_LENGTH, MAX_SKILL_NAME_LENGTH
+
+# Type alias for skill info: (name, description, scripts_count, refs_count)
+SkillInfo = Tuple[str, str, int, int]
+
+def _validate_skill_name(name: str) -> List[str]:
```
Comment (kausmeows) [inline libs/agno_infra/agno/skill/operator.py:22]: this function looks similar in logic to-
libs/agno/agno/skills/validator.py: `validate_name`

lets have a shared util if possible

File: libs/agno_infra/agno/skill/operator.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+
+        if verbose:
+            table = Table(show_header=True, header_style="bold")
+            table.add_column("Name", style="cyan")
+            table.add_column("Description")
+            table.add_column("Scripts", justify="center")
+            table.add_column("References", justify="center")
+
+            for name, desc, scripts, refs in skills_found:
+                desc_display = desc[:50] + "..." if len(desc) > 50 else desc
```
Comment (kausmeows) [inline libs/agno_infra/agno/skill/operator.py:321]: lets not truncate information if we're in verbose mode, it should show more by definition.

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR adds CLI commands for managing Agent Skills (`ag skill create`, `ag skill validate`, `ag skill list`) with proper scaffolding, validation, and listing functionality. The implementation is well-structured but suffers from significant scope creep due to multiple merges with main, making the actual changes (6 files, 738 lines) obscured by 157 total changed files.

## Architecture

The PR introduces a new CLI module for skill management:

- **`skill_cli.py`**: Typer-based CLI commands with interactive prompts
- **`operator.py`**: Core business logic for skill creation, validation, and listing
- **`templates.py`**: Template generation for SKILL.md, scripts, and references
- **`entrypoint.py`**: Integration with main agno CLI

The implementation depends on the existing `agno.skills.validator` module for validation and follows the Agent Skills specification with YAML frontmatter parsing.

## Issues

### Critical

**No test coverage** — The entire skill CLI functionality lacks tests:

```python
# No test files found for:
# - libs/agno_infra/agno/cli/skill_cli.py
# - libs/agno_infra/agno/skill/operator.py  
# - libs/agno_infra/agno/skill/templates.py
```

**Suggested fix**: Add comprehensive tests covering:
- Skill creation with various options
- Validation of valid/invalid skills
- Listing skills in different scenarios
- Error handling and edge cases
- Interactive vs non-interactive modes

**libs/agno_infra/agno/skill/operator.py:305-320** — **Import error handling is fragile** — The code assumes `agno.skills.validator` is always available:

```python
try:
    from agno.skills.validator import validate_skill_directory
except ImportError:
    console.print("[bold red]Error:[/bold red] Could not import agno.skills.validator")
    console.print("Please ensure the 'agno' package is installed:")
    console.print("  pip install agno")
    return False
```

**Suggested fix**: This should be a hard requirement with proper dependency management in `pyproject.toml`, not a runtime check.

### Warnings

**PR scope creep** — The actual feature is 6 files (738 lines) but the diff shows 157 files changed:

```
# Actual feature changes (commit a44218507):
libs/agno_infra/agno/cli/entrypoint.py    |   9 +-
libs/agno_infra/agno/cli/skill_cli.py     | 192 +++++++++++++++
libs/agno_infra/agno/skill/__init__.py      |  13 ++
libs/agno_infra/agno/skill/operator.py      | 400 +++++++++++++++++++++++++++++++
libs/agno_infra/agno/skill/templates.py     | 128 ++++++++++
libs/agno_infra/pyproject.toml              |   1 +

# But the full PR diff shows 157 files with unrelated changes
```

**Suggested fix**: Rebase the PR to remove the merge commits and keep only the skill CLI changes. Consider splitting unrelated work into separate PRs.

**libs/agno_infra/agno/skill/operator.py:145-165** — **Race condition in directory creation** — No atomic check-and-create:

```python
skill_dir = current_dir / name
if skill_dir.exists():
    console.print(f"\n[bold red]Directory already exists: {skill_dir}[/bold red]")
    return None

try:
    skill_dir.mkdir(parents=True)  # Race condition here
```

**Suggested fix**: Use `exist_ok=False` with proper error handling or catch `FileExistsError`:
```python
try:
    skill_dir.mkdir(parents=True, exist_ok=False)
except FileExistsError:
    console.print(f"\n[bold red]Directory already exists: {skill_dir}[/bold red]")
    return None
```

**libs/agno_infra/agno/skill/operator.py:88-108** — **Validation duplication** — Local validation duplicates logic from `agno.skills.validator`:

```python
def _validate_skill_name(name: str) -> List[str]:
    # This duplicates logic from agno.skills.validator._validate_name
    errors = []
    if not name or not name.strip():
        errors.append("Skill name cannot be empty")
    # ... more validation logic
```

**Suggested fix**: Use the validator from `agno.skills.validator` consistently instead of duplicating logic.

**libs/agno_infra/agno/skill/operator.py:368-420** — **YAML parsing脆弱性** — No error handling for malformed YAML:

```python
import yaml

try:
    frontmatter = yaml.safe_load(parts[1])
    if isinstance(frontmatter, dict):
        name = frontmatter.get("name", skill_path.name)
        description = frontmatter.get("description", "")
except Exception:
    pass  # Silent failure - bad for debugging
```

**Suggested fix**: Log the parsing error and provide useful feedback to the user:
```python
except yaml.YAMLError as e:
    logger.warning(f"Failed to parse YAML frontmatter: {e}")
    # Fall back to defaults
```

**libs/agno_infra/agno/cli/skill_cli.py:28-45** — **Missing command help consistency** — Some commands lack detailed help text:

```python
@skill_cli.command(short_help="Create a new skill with proper directory structure")
def create(
    # ... parameters
):
    """\b
    Create a new Agent Skill with the proper directory structure.
    """
    # But no examples section like other commands
```

**Suggested fix**: Add consistent examples section to all commands following typer best practices.

**No integration with existing skill system** — Creates new skills but doesn't integrate with `agno.skills.agent_skills`:

```python
# The CLI creates skills on disk, but there's no integration with:
# - agno.skills.agent_skills.AgentSkills
# - agno.skills.loaders.local.LocalSkillLoader  
# - Existing skill discovery and loading mechanisms
```

**Suggested fix**: Add integration tests showing how CLI-created skills work with the existing skill system.

### Suggestions

**Missing cookbook examples** — No examples showing how to use the new CLI:

```python
# Consider adding:
# cookbook/02_agents/skills/using_skill_cli.py
# cookbook/02_agents/skills/validating_skills.py
# cookbook/02_agents/skills/listing_skills.py
```

**Add `--dry-run` option** — Useful for validation without creating files:

```python
dry_run: bool = typer.Option(
    False,
    "--dry-run",
    help="Show what would be created without actually creating files.",
),
```

**Improve error messages** — Some errors could be more actionable:

```python
# Current:
console.print("[bold red]Error:[/bold red] Could not import agno.skills.validator")

# Better:
console.print("[bold red]Error:[/bold red] agno.skills.validator module not found.")
console.print("This usually means the agno package is not properly installed.")
console.print("Try: pip install -e .")
```

**Add skill update command** — Consider adding functionality to update existing skills:

```python
@skill_cli.command()
def update(
    skill_path: str = typer.Argument(..., help="Path to skill directory"),
    # ... options
):
    """Update an existing skill's metadata or structure."""
```

**Add skill export/import commands** — Useful for sharing skills:

```python
@skill_cli.command()
def export(
    skill_path: str = typer.Argument(...),
    output_file: str = typer.Option(...),
):
    """Export a skill to a portable format."""
```

**Add validation for script executability** — Check if scripts have proper shebang:

```python
# In validation:
for script in scripts_dir.iterdir():
    if script.is_file():
        content = script.read_text()
        if not content.startswith("#!"):
            errors.append(f"Script {script.name} missing shebang line")
```

## Cross-file Impact

**Public API**: Adds new CLI commands `ag skill create/validate/list`.

**Dependencies**: Depends on `agno.skills.validator` from the main agno package.

**Breaking Changes**: None.

**Integration**: Integrates with existing agno CLI but not with agno.skills loading system.

**Documentation**: No user-facing documentation or examples added.

**Test Coverage**: Critical - no tests added for new functionality.

## Assessment

⚠️ Request changes

This PR implements a useful feature with good code quality, but has several critical issues that need to be addressed:

**Major Concerns:**
1. **No test coverage** — Critical CLI functionality lacks any tests
2. **PR scope creep** — 157 files changed when only 6 files are relevant to the feature
3. **Fragile import handling** — Runtime dependency checks instead of proper dependency management
4. **No integration** — Creates skills but doesn't integrate with existing skill loading system

**Code Quality Issues:**
- Race condition in directory creation
- Duplicated validation logic
- Silent YAML parsing failures
- Missing error handling for edge cases
- No documentation or examples

**Positive Aspects:**
- Well-structured code with clear separation of concerns
- Good use of type hints and modern Python patterns
- Interactive prompts improve UX
- Comprehensive validation logic
- Follows existing CLI patterns

**Recommendations:**
1. **Rebase the PR** — Remove merge commits to show only the 6 relevant files
2. **Add comprehensive tests** — Cover all commands, edge cases, and error conditions
3. **Fix dependency management** — Make agno a proper dependency, not a runtime import
4. **Add integration examples** — Show how CLI-created skills work with the existing system
5. **Fix race conditions** — Use proper atomic operations
6. **Remove duplication** — Use agno.skills.validator consistently
7. **Add documentation** — Cookbook examples and user guide

The core implementation is solid and follows good practices, but the lack of tests and the PR's scope creep make it unsuitable for merge in its current state.
